optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,168 @@
+from collections import OrderedDict
+from typing import Any
+from typing import Dict
+from typing import Optional
+
+import scipy
+
+import optuna
+from optuna import distributions
+from optuna import logging
+from optuna._transform import _SearchSpaceTransform
+from optuna.distributions import BaseDistribution
+from optuna.samplers import BaseSampler
+from optuna.study import Study
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+_logger = logging.get_logger(__name__)
+
+_SUGGESTED_STATES = (TrialState.COMPLETE, TrialState.PRUNED)
+
+_NUMERICAL_DISTRIBUTIONS = (
+    distributions.UniformDistribution,
+    distributions.LogUniformDistribution,
+    distributions.DiscreteUniformDistribution,
+    distributions.IntUniformDistribution,
+    distributions.IntLogUniformDistribution,
+)
+
+
+class QMCSampler(BaseSampler):
+    def __init__(
+        self,
+        seed: Optional[int] = None,
+        qmc_type: str = "sobol",
+        independent_sampler: Optional[BaseSampler] = None,
+        *,
+        search_space: Optional[Dict[str, BaseDistribution]] = None,
+        warn_independent_sampling: bool = True,
+    ) -> None:
+
+        self._seed = seed
+        self._independent_sampler = independent_sampler or optuna.samplers.RandomSampler(seed=seed)
+        self._qmc_type = qmc_type
+        self._qmc_engine = None
+        # TODO(kstoneriv3): make sure that search_space is either None or valid search space.
+        self._initial_search_space = search_space
+        self._warn_independent_sampling = warn_independent_sampling
+
+    def infer_relative_search_space(
```

Comment:
The logic to determine the search space is very different from other samplers. In this `QMCSampler`, the search space is wholly determined after the first trial finish and cannot be changed in the subsequent trials. So, `QMCSampler` does not support the "conditional search space". We need to note that or change the logic to allow varying search space if possible.

Reply:
OK, I will note this point. However, I do not understand what varying search space means. If the search space of `QMCSampler` needs to contain a parameter that is not suggested in the first trial, we can provide them at the `__init__(...)` I guess. 
(As a side remark:) the QMC sequence is a sequence that fills the d-dimensional hypercube "more uniformly" (in some sense) than a random sequence. Thus, we need to determine the dimension d (the number of parameters) when we construct the `QMCEngine` of `scipy` so that we know the number of dimensions before starting the sampling.

Reply:
Thanks for the comment. In Optuna, the user can define a conditional search space: a parameter sampled in the first trial may not be sampled in the second trial. If QMC is not suitable for such cases, it would be helpful to document it.

Reply:
OK, I wrote the docstrings so could you take a look and see if this point is well documented? However, I still feel I did not understand your point so it would be helpful if you could add some comments at the docstrings.

Reply:
I think the point was that relative samplers in Optuna usually compute search space intersections in _each trial_  (using `IntersectionSearchSpace ` in e.g.  [`_cmaes.py`](https://github.com/optuna/optuna/blob/master/optuna/samplers/_cmaes.py)). This naturally gradually narrows down the search space if parameters are conditioned/differ between trials. This would not be an issue if the sampler is stateless but I'm not sure if it's difficult.

Reply:
So, I changed the design so that the sampler will be more stateless. Though the `infer_relative_search_space` returns the same search space all the time, the `sample_relative` function can now handle arbitrary `search_space`. When unseen `search_space` is given to the `sample_relative`, it samples from a new QMC sequence.